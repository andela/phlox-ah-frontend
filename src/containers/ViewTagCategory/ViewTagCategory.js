import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Row, Col } from 'react-materialize';

import { ArticleCard } from '../../components/ArticleCard/ArticleCard';
import Sidebar from '../Sidebar/Sidebar';
import { getOneTag } from '../../requests/TagRequests';
import { getCategoryArticle } from '../../requests/CategoryRequests';
import { getArticles } from '../../requests/ArticleRequests';
import './ViewTagCategory.scss';


/**
 * @class ViewTagCategory
 * @extends {Component}
 */
class ViewTagCategory extends Component {
  /**
   *Creates an instance of Home.
   * @param {object} props
   * @memberof ViewTagCategory
   */
  constructor(props) {
    super();
    this.state = {
      success: false,
      tag: {},
      tagName: '',
      categorySelected: false,
      categories: []
    };

    this.getTag = this.getTag.bind(this);
  }

  /**
   * @description - This method runs whenever redux state changes
   * @returns {object} state
   * @param {object} props
   * @param {object} state
   * @memberof ViewTagCategory
   */
  static getDerivedStateFromProps(props, state) {
    const { category, name } = props.match.params;
    if (name !== state.tagName) {
      props.getOneTag(name);
    }
    return {
      success: props.success,
      tag: props.tag,
      tagName: name,
      categories: props.categories,
      categorySelected: Boolean(category)
    };
  }

  /**
   *
   * @returns {func} tag
   * @memberof ViewTagCategory
   */
  componentDidMount() {
    const { category, name } = this.props.match.params;
    if (category) {
      this.setState({ categorySelected: true });
      this.props.getCategoryArticle(category);
    } else {
      this.props.getOneTag(name);
    }
    this.props.getArticles();
  }

  /**
   *
   *
   * @param {*} name
   * @returns {func} tag
   * @memberof ViewTagCategory
   */
  getTag(name) {
    this.props.getOneTag(name);
  }

  /**
   *
   *
   * @returns {jsx} - jsx
   * @memberof ViewTagCategory
   */
  showTagArticles() {
    if (this.state.success && this.state.tag) {
      return this.state.tag.Articles.map((article, index) => {
        let username = '';
        if (article.User) {
          username = article.User.username;
        }
        return (<Col s={12} m={12}
        l={12} xl={6} key={article.id}>
        <ArticleCard size="medium" pic={article.imgUrl} title={article.title} description={article.description} createdAt={article.createdAt} author={username} slug={article.slug} tags={article.Tags} />
        </Col>);
      });
    }
    return (<h5 className="center no-tags">There are no articles under this tag</h5>);
  }

  /**
   *
   *
   * @returns {jsx} - jsx
   * @memberof ViewTagCategory
   */
  showCategoryArticles() {
    if (this.state.success && this.state.categories.length) {
      return this.state.categories.map((article, index) => {
        let username = '';
        if (article.User) {
          username = article.User.username;
        }
        return (<Col s={12} m={12}
        l={12} xl={6} key={article.id}>
        <ArticleCard size="medium" pic={article.imgUrl} title={article.title} description={article.description} createdAt={article.createdAt} author={username} slug={article.slug} tags={article.Tags} />
        </Col>);
      });
    }
    return (<h5 className="center no-tags">There are no articles under this category</h5>);
  }

  /**
   *
   * @description - This method renders the jsx for this component
   * @returns {jsx} - jsx
   * @memberof ViewTagCategory
   */
  render() {
    return (
      <div>
         <main>
          <div className="site-container">
          <Row>
              <Col s={12} m={12} l={8} xl={9} className="column-1">
                <Row>
                  <h5 className="center capitalize">{this.state.categorySelected ? 'Category' : 'Tag'}: { this.state.categorySelected ? this.props.match.params.category : this.props.match.params.name }</h5>
                  { this.state.categorySelected ? this.showCategoryArticles() : this.showTagArticles() }
                </Row>
              </Col>
              <Col s={12} m={9} l={4} xl={3} className="sidebar">
                <Sidebar />
              </Col>
          </Row>
          </div>
        </main>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  articles: state.article.articles,
  tag: state.tags.tag,
  categories: state.category.categoryArticle,
  success: state.tags.success,
  loading: state.category.loading
});

ViewTagCategory.propTypes = {
  articles: PropTypes.array,
  getOneTag: PropTypes.func,
  getArticles: PropTypes.func,
  getCategoryArticle: PropTypes.func,
  match: PropTypes.object,
  success: PropTypes.bool,
  loading: PropTypes.bool,
  tag: PropTypes.object
};

export default connect(mapStateToProps, {
  getOneTag, getArticles, getCategoryArticle
})(ViewTagCategory);
