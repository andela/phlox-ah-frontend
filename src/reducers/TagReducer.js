import { ALL_TAGS, ONE_TAG } from '../actionTypes';
import { asyncActionName } from '../util/AsyncUtil';

const initialState = {
  tag: {},
  tags: [],
  message: null,
  error: null,
  loading: false,
  failure: false,
  success: false
};

const TagReducer = (state = initialState, action) => {
  switch (action.type) {
    case asyncActionName(ALL_TAGS).success:
      return {
        ...state,
        tags: action.payload.tags,
        message: action.payload.message
      };
    case asyncActionName(ALL_TAGS).failure:
      return {
        ...state, error: action.payload.status, message: action.payload.error
      };
    case asyncActionName(ONE_TAG).loading:
      return { ...state, loading: action.payload };
    case asyncActionName(ONE_TAG).success:
      return {
        ...state, tag: action.payload, success: true
      };
    case asyncActionName(ONE_TAG).failure:
      return {
        ...state,
        error: action.payload.status,
        failure: true,
      };
    default:
      return state;
  }
};

export default TagReducer;
