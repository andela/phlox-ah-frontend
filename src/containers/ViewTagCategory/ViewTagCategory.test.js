import React from 'react';
import { shallow } from 'enzyme';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import ViewTagCategory from './ViewTagCategory';
import TagReducer from '../../reducers/TagReducer';

const mockStore = configureMockStore([thunk]);
const store = mockStore({
  success: true,
  article: {
    articles: []
  },
  tags: {
    tag: { articles: [{ title: 'first title' }] }
  },
  category: {
    categoryArticle: []
  },
  getOneTag: jest.fn()

});

let component;
let myComponent;

const props = {
  match: {
    url: '/tags/sports',
    params: {
      name: ''
    }
  }
};

describe('<ViewTagCategory/>', () => {
  beforeEach(() => {
    component = shallow(
      <Provider store={store}>
        <ViewTagCategory {...props}/>
      </Provider>
    );
    myComponent = component.dive({ context: { store } }).dive();
  });
  it('should render without throwing an error', () => {
    expect(component).toMatchSnapshot();
  });
  it('should have a site-container class', () => {
    expect(myComponent.find('.site-container').exists()).toBe(true);
  });
  it('should have a sidebar class', () => {
    expect(myComponent.find('.sidebar').exists()).toBe(true);
  });
  it('should have a h5', () => {
    expect(myComponent.find('h5.center').exists()).toBe(true);
  });
  it('should have function showTagArticles', () => {
    expect(myComponent.instance().showTagArticles).toBeDefined();
  });
  it('should have function getTag', () => {
    expect(myComponent.instance().getTag).toBeDefined();
  });
  it('should have tag state as object', () => {
    expect(myComponent.instance().state.tag).toMatchObject({});
  });
  it('should have tagName state as string', () => {
    expect(myComponent.instance().state.tagName).toEqual('');
  });
  it('should have articles props', () => {
    expect(myComponent.instance().props.match.url).toEqual('/tags/sports');
  });
  it('should dispatch an action to get all tags', () => {
    const payload = {
      tags: [],
      message: 'Tags retrieved successfullys',
    };
    const action = {
      type: 'ALL_TAGS_SUCCESS',
      payload
    };
    const newState = TagReducer({}, action);
    expect(newState).toEqual({
      tags: payload.tags,
      message: payload.message,
    });
  });
});
