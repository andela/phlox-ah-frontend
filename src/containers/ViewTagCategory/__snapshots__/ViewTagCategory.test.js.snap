// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ViewTagCategory/> should render without throwing an error 1`] = `
<Connect(ViewTagCategory)
  match={
    Object {
      "params": Object {
        "name": "",
      },
      "url": "/tags/sports",
    }
  }
/>
`;
